include/power/pm_policy.h

```diff
@@ -0,0 +1,183 @@
+ * macro in the beginning of the declaration and be statically defined
+ * using @c PM_POLICY_DEFINE.
+ */
+#define PM_POLICY_BASE \
+	struct { \
+		const char *name; \
+		const size_t len; \
+		enum power_states (*next_state_get)(struct pm_policy *policy,\
+						    int32_t ticks); \
+		bool (*low_power_devices_get)(struct pm_policy *policy, \
+					      enum power_states state); \
+	}
+
+/**
+ * @brief PM Policy definition
+ *
+ * Store information about a power management policy.
+ */
+struct pm_policy {
+	PM_POLICY_BASE;
+};
+
+/**
+ * @brief Statically define and register a power management policy
+ *
+ * The policy can be accessed outside the module where it is defined using:
+ *
+ * @code extern struct <policy_type> <policy_name>; @endcode
+ *
+ * @param policy_type The pm_policy struct used
+ * @param policy_name The name of this policy
+ * @param policy_next_state_get Callback to get the next PM state
+ * @param policy_low_power_devices_get Callback to determine whether to put
+ *        devices in low power state, given the system PM state
+ */
+#define PM_POLICY_DEFINE(policy_type, policy_name, policy_next_state_get, \
+			 policy_low_power_devices_get, ...)		\
+	Z_STRUCT_SECTION_ITERABLE_ALTERNATE(pm_policy, policy_type,	\
+					    policy_name) = {		\
+		.name = STRINGIFY(policy_name),			\
+		.len = sizeof(struct policy_type),			\
+		.next_state_get = policy_next_state_get,		\
+		.low_power_devices_get = policy_low_power_devices_get,	\
+		__VA_ARGS__						\
+	}
+
+#define PM_POLICY_FOREACH(struct_type, iterator)			\
+	extern struct struct_type _CONCAT(_##struct_type, _list_start)[]; \
+	extern struct struct_type _CONCAT(_##struct_type, _list_end)[]; \
+	for (struct struct_type *iterator =				\
+		     _CONCAT(_##struct_type, _list_start);		\
+	     ({ __ASSERT(iterator <= _CONCAT(_##struct_type, _list_end), \
+			 "unexpected list end location");		\
+		     iterator < _CONCAT(_##struct_type, _list_end); }); \
+	     iterator = (struct struct_type *)				\
+		     ((uint8_t *)iterator + iterator->len))
+
+/**
+ * @brief Function to create device PM list
+ */
+void sys_pm_create_device_list(void);
+
+/**
+ * @brief Function to suspend the devices in PM device list
+ */
+int sys_pm_suspend_devices(void);
+
+/**
+ * @brief Function to put the devices in PM device list in low power state
+ */
+int sys_pm_low_power_devices(void);
+
+/**
+ * @brief Function to force suspend the devices in PM device list
+ */
+int sys_pm_force_suspend_devices(void);
+
+/**
+ * @brief Function to resume the devices in PM device list
+ */
+void sys_pm_resume_devices(void);
+
+/**
+ * @brief Function to get the next PM state based on the ticks
+ */
+static inline enum power_states sys_pm_policy_next_state(int32_t ticks)
+{
+	extern struct pm_policy *sys_pm_policy;
+
+	if (sys_pm_policy == NULL) {
+		return -EINVAL;
+	}
+
+	if (sys_pm_policy->next_state_get == NULL) {
+		/* Should it be -ENOSYS ? */
```

Comment:
-ENOTSUP seems fine to me

Reply:
ok

Reply:
Initially I had used ENOSYS, but then in checkpatch.pl we have "ENOSYS means 'invalid syscall nr' and nothing else\n", that is not the case here and I saw some places using ENOTSUP. I think ENOSYS is better, but I tried to be more consistent with the rest of the code base.

Reply:
The issue with checkpatch is that it's entirely Linux centric, which generates troubles as this one sometimes.